File: code/julia/rest/client/src/TaskClient.jl
module TaskClient

using HTTP
using JSON3

struct Task
    id::String
    title::String
    description::Union{String, Nothing}
    status::String
    priority::String
    tags::Vector{String}
    assigned_to::Union{String, Nothing}
    created_at::Float64
    updated_at::Float64
end

struct Client
    base_url::String
end

function list_tasks(client::Client; status=nothing, assigned_to=nothing)
    url = "$(client.base_url)/api/tasks"
    params = Dict{String, String}()
    
    if !isnothing(status)
        params["status"] = status
    end
    if !isnothing(assigned_to)
        params["assigned_to"] = assigned_to
    end
    
    if !isempty(params)
        query_string = join(["$k=$v" for (k, v) in params], "&")
        url = "$url?$query_string"
    end
    
    response = HTTP.get(url)
    
    if response.status == 200
        data = JSON3.read(String(response.body))
        return data.tasks
    else
        return nothing
    end
end

function get_task(client::Client, id::String)
    url = "$(client.base_url)/api/tasks/$id"
    response = HTTP.get(url)
    
    if response.status == 200
        return JSON3.read(String(response.body))
    else
        return nothing
    end
end

function create_task(client::Client, title::String; description=nothing, priority="medium", tags=String[], assigned_to=nothing)
    url = "$(client.base_url)/api/tasks"
    
    body = Dict(
        "title" => title,
        "priority" => priority,
        "tags" => tags
    )
    
    if !isnothing(description)
        body["description"] = description
    end
    if !isnothing(assigned_to)
        body["assigned_to"] = assigned_to
    end
    
    response = HTTP.post(
        url,
        ["Content-Type" => "application/json"],
        JSON3.write(body)
    )
    
    if response.status == 201
        return JSON3.read(String(response.body))
    else
        return nothing
    end
end

function update_task(client::Client, id::String; title=nothing, description=nothing, status=nothing, priority=nothing, tags=nothing, assigned_to=nothing)
    url = "$(client.base_url)/api/tasks/$id"
    
    body = Dict{String, Any}()
    
    if !isnothing(title)
        body["title"] = title
    end
    if !isnothing(description)
        body["description"] = description
    end
    if !isnothing(status)
        body["status"] = status
    end
    if !isnothing(priority)
        body["priority"] = priority
    end
    if !isnothing(tags)
        body["tags"] = tags
    end
    if !isnothing(assigned_to)
        body["assigned_to"] = assigned_to
    end
    
    response = HTTP.put(
        url,
        ["Content-Type" => "application/json"],
        JSON3.write(body)
    )
    
    if response.status == 200
        return JSON3.read(String(response.body))
    else
        return nothing
    end
end

function update_task_status(client::Client, id::String, status::String)
    url = "$(client.base_url)/api/tasks/$id/status"
    
    body = Dict("status" => status)
    
    response = HTTP.patch(
        url,
        ["Content-Type" => "application/json"],
        JSON3.write(body)
    )
    
    if response.status == 200
        return JSON3.read(String(response.body))
    else
        return nothing
    end
end

function delete_task(client::Client, id::String)
    url = "$(client.base_url)/api/tasks/$id"
    response = HTTP.delete(url)
    
    return response.status == 204
end

function print_banner()
    println("╔════════════════════════════════════════════════╗")
    println("║       Julia Task Management REST Client        ║")
    println("║            Testing API Operations              ║")
    println("╚════════════════════════════════════════════════╝")
    println()
end

function run_demo()
    client = Client("http://localhost:8080")
    
    # 1. List all tasks
    println("1. Listing all tasks...")
    tasks = list_tasks(client)
    if !isnothing(tasks)
        println("   Found $(length(tasks)) tasks")
        for task in tasks
            println("   - [$(task.id)] $(task.title) ($(task.status))")
        end
    else
        println("   Error: Failed to list tasks")
    end
    
    # 2. Create a new task
    println("\n2. Creating a new task...")
    new_task = create_task(
        client,
        "Learn Julia multiple dispatch",
        description="Master Julia's powerful type system and multiple dispatch",
        priority="high",
        tags=["julia", "dispatch", "types"],
        assigned_to="julia-team"
    )
    
    if !isnothing(new_task)
        println("   Created task: $(new_task.title)")
        println("   ID: $(new_task.id)")
        println("   Priority: $(new_task.priority)")
        println("   Tags: $(join(new_task.tags, ", "))")
        
        task_id = new_task.id
        
        # 3. Get task details
        println("\n3. Getting task details...")
        task_detail = get_task(client, task_id)
        if !isnothing(task_detail)
            println("   Title: $(task_detail.title)")
            if !isnothing(task_detail.description)
                println("   Description: $(task_detail.description)")
            end
            println("   Status: $(task_detail.status)")
            if !isnothing(task_detail.assigned_to)
                println("   Assigned to: $(task_detail.assigned_to)")
            end
        else
            println("   Error: Failed to get task")
        end
        
        # 4. Update task status
        println("\n4. Updating task status to 'in-progress'...")
        status_update = update_task_status(client, task_id, "in-progress")
        if !isnothing(status_update)
            println("   Updated status to: $(status_update.status)")
        else
            println("   Error: Failed to update status")
        end
        
        # 5. Update task details
        println("\n5. Updating task details...")
        task_update = update_task(
            client,
            task_id,
            title="Master Julia's metaprogramming",
            priority="urgent"
        )
        if !isnothing(task_update)
            println("   Updated title: $(task_update.title)")
            println("   Updated priority: $(task_update.priority)")
        else
            println("   Error: Failed to update task")
        end
        
        # 6. Filter tasks by status
        println("\n6. Filtering tasks by status...")
        in_progress_tasks = list_tasks(client, status="in-progress")
        if !isnothing(in_progress_tasks)
            println("   Found $(length(in_progress_tasks)) in-progress tasks")
            for task in in_progress_tasks
                println("   - $(task.title)")
            end
        else
            println("   Error: Failed to filter tasks")
        end
        
        # 7. Delete the task
        println("\n7. Deleting the task...")
        if delete_task(client, task_id)
            println("   Task deleted successfully")
        else
            println("   Error: Failed to delete task")
        end
        
        # 8. Verify deletion
        println("\n8. Verifying deletion...")
        verify_task = get_task(client, task_id)
        if isnothing(verify_task)
            println("   Task not found (as expected)")
        else
            println("   Error: Task still exists")
        end
    else
        println("   Error: Failed to create task")
    end
    
    println("\n✅ Demo completed successfully!")
end

end # module

File: code/nim/rest/client/src/task_client.nim
import puppy
import json
import strformat
import options
import os
import strutils

type
  TaskClient* = object
    baseUrl: string
  
  Task* = object
    id*: string
    title*: string
    description*: Option[string]
    status*: string
    priority*: string
    tags*: seq[string]
    assignedTo*: Option[string]
    createdAt*: float
    updatedAt*: float

proc newTaskClient*(baseUrl: string = "http://localhost:8080"): TaskClient =
  TaskClient(baseUrl: baseUrl)

proc listTasks*(client: TaskClient, status: string = "", assignedTo: string = ""): seq[Task] =
  var url = client.baseUrl & "/api/tasks"
  var params: seq[string] = @[]
  
  if status != "":
    params.add("status=" & status)
  if assignedTo != "":
    params.add("assigned_to=" & assignedTo)
  
  if params.len > 0:
    url = url & "?" & params.join("&")
  
  let response = fetch(url)
  if response.code == 200:
    let data = parseJson(response.body)
    result = @[]
    for taskJson in data["tasks"]:
      var task = Task(
        id: taskJson["id"].getStr(),
        title: taskJson["title"].getStr(),
        status: taskJson["status"].getStr(),
        priority: taskJson["priority"].getStr(),
        tags: taskJson["tags"].getElems().mapIt(it.getStr()),
        createdAt: taskJson["created_at"].getFloat(),
        updatedAt: taskJson["updated_at"].getFloat()
      )
      
      if taskJson.hasKey("description"):
        task.description = some(taskJson["description"].getStr())
      else:
        task.description = none(string)
      
      if taskJson.hasKey("assigned_to"):
        task.assignedTo = some(taskJson["assigned_to"].getStr())
      else:
        task.assignedTo = none(string)
      
      result.add(task)
  else:
    result = @[]

proc getTask*(client: TaskClient, id: string): Option[Task] =
  let url = client.baseUrl & "/api/tasks/" & id
  let response = fetch(url)
  
  if response.code == 200:
    let taskJson = parseJson(response.body)
    var task = Task(
      id: taskJson["id"].getStr(),
      title: taskJson["title"].getStr(),
      status: taskJson["status"].getStr(),
      priority: taskJson["priority"].getStr(),
      tags: taskJson["tags"].getElems().mapIt(it.getStr()),
      createdAt: taskJson["created_at"].getFloat(),
      updatedAt: taskJson["updated_at"].getFloat()
    )
    
    if taskJson.hasKey("description"):
      task.description = some(taskJson["description"].getStr())
    else:
      task.description = none(string)
    
    if taskJson.hasKey("assigned_to"):
      task.assignedTo = some(taskJson["assigned_to"].getStr())
    else:
      task.assignedTo = none(string)
    
    result = some(task)
  else:
    result = none(Task)

proc createTask*(client: TaskClient, title: string, description: string = "", 
                priority: string = "medium", tags: seq[string] = @[], 
                assignedTo: string = ""): Option[Task] =
  let url = client.baseUrl & "/api/tasks"
  var body = %* {
    "title": title,
    "priority": priority,
    "tags": tags
  }
  
  if description != "":
    body["description"] = %description
  
  if assignedTo != "":
    body["assigned_to"] = %assignedTo
  
  let response = fetch(
    url,
    verb = "post",
    headers = @[Header(key: "Content-Type", value: "application/json")],
    body = $body
  )
  
  if response.code == 201:
    let taskJson = parseJson(response.body)
    var task = Task(
      id: taskJson["id"].getStr(),
      title: taskJson["title"].getStr(),
      status: taskJson["status"].getStr(),
      priority: taskJson["priority"].getStr(),
      tags: taskJson["tags"].getElems().mapIt(it.getStr()),
      createdAt: taskJson["created_at"].getFloat(),
      updatedAt: taskJson["updated_at"].getFloat()
    )
    
    if taskJson.hasKey("description"):
      task.description = some(taskJson["description"].getStr())
    else:
      task.description = none(string)
    
    if taskJson.hasKey("assigned_to"):
      task.assignedTo = some(taskJson["assigned_to"].getStr())
    else:
      task.assignedTo = none(string)
    
    result = some(task)
  else:
    result = none(Task)

proc updateTask*(client: TaskClient, id: string, title: string = "", 
                description: string = "", status: string = "", 
                priority: string = "", tags: seq[string] = @[],
                assignedTo: string = ""): Option[Task] =
  let url = client.baseUrl & "/api/tasks/" & id
  var body = newJObject()
  
  if title != "":
    body["title"] = %title
  if description != "":
    body["description"] = %description
  if status != "":
    body["status"] = %status
  if priority != "":
    body["priority"] = %priority
  if tags.len > 0:
    body["tags"] = %tags
  if assignedTo != "":
    body["assigned_to"] = %assignedTo
  
  let response = fetch(
    url,
    verb = "put",
    headers = @[Header(key: "Content-Type", value: "application/json")],
    body = $body
  )
  
  if response.code == 200:
    let taskJson = parseJson(response.body)
    var task = Task(
      id: taskJson["id"].getStr(),
      title: taskJson["title"].getStr(),
      status: taskJson["status"].getStr(),
      priority: taskJson["priority"].getStr(),
      tags: taskJson["tags"].getElems().mapIt(it.getStr()),
      createdAt: taskJson["created_at"].getFloat(),
      updatedAt: taskJson["updated_at"].getFloat()
    )
    
    if taskJson.hasKey("description"):
      task.description = some(taskJson["description"].getStr())
    else:
      task.description = none(string)
    
    if taskJson.hasKey("assigned_to"):
      task.assignedTo = some(taskJson["assigned_to"].getStr())
    else:
      task.assignedTo = none(string)
    
    result = some(task)
  else:
    result = none(Task)

proc updateTaskStatus*(client: TaskClient, id: string, status: string): Option[Task] =
  let url = client.baseUrl & "/api/tasks/" & id & "/status"
  let body = %* {"status": status}
  
  let response = fetch(
    url,
    verb = "patch",
    headers = @[Header(key: "Content-Type", value: "application/json")],
    body = $body
  )
  
  if response.code == 200:
    let taskJson = parseJson(response.body)
    var task = Task(
      id: taskJson["id"].getStr(),
      title: taskJson["title"].getStr(),
      status: taskJson["status"].getStr(),
      priority: taskJson["priority"].getStr(),
      tags: taskJson["tags"].getElems().mapIt(it.getStr()),
      createdAt: taskJson["created_at"].getFloat(),
      updatedAt: taskJson["updated_at"].getFloat()
    )
    
    if taskJson.hasKey("description"):
      task.description = some(taskJson["description"].getStr())
    else:
      task.description = none(string)
    
    if taskJson.hasKey("assigned_to"):
      task.assignedTo = some(taskJson["assigned_to"].getStr())
    else:
      task.assignedTo = none(string)
    
    result = some(task)
  else:
    result = none(Task)

proc deleteTask*(client: TaskClient, id: string): bool =
  let url = client.baseUrl & "/api/tasks/" & id
  let response = fetch(url, verb = "delete")
  result = response.code == 204

proc printBanner() =
  echo "╔════════════════════════════════════════════════╗"
  echo "║        Nim Task Management REST Client         ║"
  echo "║            Testing API Operations              ║"
  echo "╚════════════════════════════════════════════════╝"
  echo ""

proc runDemo() =
  let client = newTaskClient()
  
  # 1. List all tasks
  echo "1. Listing all tasks..."
  let tasks = client.listTasks()
  echo fmt"   Found {tasks.len} tasks"
  for task in tasks:
    echo fmt"   - [{task.id}] {task.title} ({task.status})"
  
  # 2. Create a new task
  echo "\n2. Creating a new task..."
  let newTaskOpt = client.createTask(
    title = "Learn Nim metaprogramming",
    description = "Master macros and compile-time execution",
    priority = "high",
    tags = @["nim", "metaprogramming", "macros"],
    assignedTo = "nim-team"
  )
  
  if newTaskOpt.isSome:
    let task = newTaskOpt.get()
    echo fmt"   Created task: {task.title}"
    echo fmt"   ID: {task.id}"
    echo fmt"   Priority: {task.priority}"
    echo fmt"   Tags: {task.tags.join(\", \")}"
    
    # 3. Get task details
    echo "\n3. Getting task details..."
    let taskDetailOpt = client.getTask(task.id)
    if taskDetailOpt.isSome:
      let taskDetail = taskDetailOpt.get()
      echo fmt"   Title: {taskDetail.title}"
      if taskDetail.description.isSome:
        echo fmt"   Description: {taskDetail.description.get()}"
      echo fmt"   Status: {taskDetail.status}"
      if taskDetail.assignedTo.isSome:
        echo fmt"   Assigned to: {taskDetail.assignedTo.get()}"
    
    # 4. Update task status
    echo "\n4. Updating task status to 'in-progress'..."
    let statusUpdateOpt = client.updateTaskStatus(task.id, "in-progress")
    if statusUpdateOpt.isSome:
      echo fmt"   Updated status to: {statusUpdateOpt.get().status}"
    
    # 5. Update task details
    echo "\n5. Updating task details..."
    let updateOpt = client.updateTask(
      task.id,
      title = "Master Nim async and parallelism",
      priority = "urgent"
    )
    if updateOpt.isSome:
      let updated = updateOpt.get()
      echo fmt"   Updated title: {updated.title}"
      echo fmt"   Updated priority: {updated.priority}"
    
    # 6. Filter tasks by status
    echo "\n6. Filtering tasks by status..."
    let inProgressTasks = client.listTasks(status = "in-progress")
    echo fmt"   Found {inProgressTasks.len} in-progress tasks"
    for t in inProgressTasks:
      echo fmt"   - {t.title}"
    
    # 7. Delete the task
    echo "\n7. Deleting the task..."
    if client.deleteTask(task.id):
      echo "   Task deleted successfully"
    else:
      echo "   Error: Failed to delete task"
    
    # 8. Verify deletion
    echo "\n8. Verifying deletion..."
    let verifyOpt = client.getTask(task.id)
    if verifyOpt.isNone:
      echo "   Task not found (as expected)"
    else:
      echo "   Error: Task still exists"
  else:
    echo "   Error: Failed to create task"
  
  echo "\n✅ Demo completed successfully!"

when isMainModule:
  printBanner()
  sleep(1000)  # Give server time to be ready
  runDemo()